### PR TITLE
Add Goroutine leak detection with unit tests

### PR DESCRIPTION
**Pull request by OhmSpectator**
This pull request introduces a goroutine leak detection system in the watcher component, along with a comprehensive test suite to validate the accuracy and reliability of leak identification. Additionally, a minor update ensures stack dumps work seamlessly within `watcher` by verifying that the required directory is created as needed.

Key Changes:
* Goroutine Leak Detection in watcher
The new detection system monitors goroutine counts to identify potential leaks by combining a threshold check for immediate spikes with trend-based analysis. This approach smooths data over time and uses dynamically adjusted thresholds to flag gradual increases, potentially indicating leaks.

* Testing for Leak Scenarios
A suite of tests simulates various scenarios including growth patterns, spikes, gradual leaks, and edge cases to ensure the detector accurately distinguishes between normal variations and true leaks.

TODO:
- [x] Test in a known leak scenario (in progress)
- [x] Make the detector parameters configurable
- [x] Add the documentation

### CODE REVIEW

File: pkg/pillar/cmd/watcher/watcher.go [hunk 1]
```diff
@@ -156,6 +157,101 @@ func handleMemoryPressureEvents() {
...
+		if len(stats) > maxEntries {
+			stats = stats[1:]
+		}
+		// Check if there are any leaks
+		if len(stats) > 60 && detectGoroutineLeaks(stats) {
+			// Count the number of goroutines that were created in the last 1 hour
+			numGoroutinesHourAgo := stats[len(stats)-60]
+			leakCount := numGoroutines - numGoroutinesHourAgo
+			log.Warnf("Goroutine leak detected. Number of goroutines created in the last 1 hour: %d", leakCount)
+			// TODO, call function similar to dumpStacks() in agentlog.go
```
Comment (eriknordmark) [inline pkg/pillar/cmd/watcher/watcher.go:250]: Could the dumpStacks output be processed to try to detect the stack traces for the potential leaks?
Diffing against a saved dumpstacks from an hour earlier might generate a bunch of noise, but if there are brand new goroutine numbers and there is one (or more) common stack traces in those newbies, it would be great if we could log those common stack traces.

Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher.go]: My initial approach is to detect potential leaks by identifying a stable growth in the number of goroutines over time. I've implemented logic to minimize false positives caused by fluctuations and temporary spikes, and synthetic tests indicate this approach works well.

In the future, I plan to add stack trace analysis. However, this will be a more complex implementation, so I’d like to start with a simpler approach based on monitoring goroutine counts.

Comment (eriknordmark) [reply pkg/pillar/cmd/watcher/watcher.go]: Ok

**Review by eriknordmark — APPROVED** [review]
LGTM

**Review by deitch — COMMENTED** [review]
This is impressive. I liked the advanced thought in this. A few questions:

1. Is it possible to change the constants for timing and thresholds to variables, with the defaults being constants, and they can be overridden via eve config? There may be legitimate scenarios where what the constants consider a problem is perfectly normal operation. We don't want to have to upgrade EVE on devices to change them when a config would do.
2. Can we document this?

**Comment by OhmSpectator** [comment]
> This is impressive. I liked the advanced thought in this. A few questions:
> 
> 1. Is it possible to change the constants for timing and thresholds to variables, with the defaults being constants, and they can be overridden via eve config? There may be legitimate scenarios where what the constants consider a problem is perfectly normal operation. We don't want to have to upgrade EVE on devices to change them when a config would do.
> 2. Can we document this?

Yep, I want to do both. I should have done it before converting from Draft =)

**Comment by OhmSpectator** [comment]
> This is impressive. I liked the advanced thought in this. A few questions:
> 
> 1. Is it possible to change the constants for timing and thresholds to variables, with the defaults being constants, and they can be overridden via eve config? There may be legitimate scenarios where what the constants consider a problem is perfectly normal operation. We don't want to have to upgrade EVE on devices to change them when a config would do.
> 2. Can we document this?

Done!

File: docs/CONFIG-PROPERTIES.md [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # EVE Runtime Configuration Properties
 
 | Name | Type | Default | Description |
-| ---- | ---- | ------- | ----------- |
+| ---- | ---- | ------ | ----------- |
```
Comment (eriknordmark) [inline docs/CONFIG-PROPERTIES.md:4]: Is this an intentional change?

File: docs/CONFIG-PROPERTIES.md [hunk 3]
```diff
@@ -64,7 +64,11 @@
...
-
+| goroutine.leak.detection.threshold | integer | 5000 | Amount of goroutines, reaching which will trigger leak detection regardless of growth rate. |
+| goroutine.leak.detection.check.interval.minutes | integer (minutes) | 1 | Interval in minutes between the measurements of the goroutine count. |
+| goroutine.leak.detection.check.window.minutes | integer (minutes) | 10 | Interval in minutes for which the leak analysis is performed. It should contain at least 10 measurements, so no less than 10 × goroutine.leak.detection.check.interval.minutes. |
+| goroutine.leak.detection.keep.stats.hours | integer (hours) | 24 | Amount of hours to keep the stats for leak detection. We keep more stats than the check window to be able to react to settings with a bigger check window via configuration. |
+| goroutine.leak.detection.cooldown.minutes | integer (minutes) | 5 | Cooldown period in minutes after the leak detection is triggered. During this period, no stack traces are collected; only warning messages are logged. |
 In addition, there can be per-agent settings.
```
Comment (eriknordmark) [inline docs/CONFIG-PROPERTIES.md:72]: Missing blank line? Or does the markdown formatter leave some blank space?

File: pkg/pillar/cmd/watcher/watcher.go [hunk 4]
```diff
@@ -268,11 +529,17 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject, ar
 	// Handle memory pressure events by calling GC explicitly
 	go handleMemoryPressureEvents()
 
+	// Set the goroutine leak detection parameters to default values
+	ctx.GRLDParams.Set(5000, 1*time.Minute, 10*time.Minute, 24*time.Hour, 5*time.Minute)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/watcher/watcher.go:533]: You don't need this; a readGlobalGoroutineLeakDetectionParams() is sufficient.

You have already waited for GCinitialized, so all of the ConfigItems are known.
And the upgradeconverter (or is it zedagent) ensures that all new ConfigItems are filled in with their default values before the ConfigItems are published.

File: docs/CONFIG-PROPERTIES.md (hunk 2 above)
Comment (OhmSpectator) [reply docs/CONFIG-PROPERTIES.md:4]: Nope, will remove now )

File: docs/CONFIG-PROPERTIES.md (hunk 3 above)
Comment (OhmSpectator) [reply docs/CONFIG-PROPERTIES.md:72]: It was already like that... But I can add a blank line there. It also looks strange to me.

File: pkg/pillar/cmd/watcher/watcher.go (hunk 4 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher.go]: Removed!

File: pkg/pillar/docs/watcher.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
+# Watcher
+
+The **watcher** component is the part of our system responsible for
+monitoring and managing system resources to ensure optimal performance and
+stability. It oversees various aspects such as memory usage, disk space, and the
+number of active goroutines. By keeping track of these metrics, the watcher can
+detect anomalies, trigger garbage collection, and alert us to potential issues
+like resource exhaustion or leaks.
+
+## Manual Garbage Collection
```
Comment (eriknordmark) [inline pkg/pillar/docs/watcher.md:10]: In addition to these recent additions, it would make sense to also document the original logging around HostMemory and DiskUsage. Maybe @gkodali-zededa can propose a paragraph or two?

File: pkg/pillar/cmd/watcher/watcher.go [hunk 6]
```diff
@@ -268,11 +529,14 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject, ar
...
 	go handleMemoryPressureEvents()
 
+	go goroutinesMonitor(&ctx)
+
 	for {
 		select {
 		case change := <-subGlobalConfig.MsgChan():
 			subGlobalConfig.ProcessChange(change)
 			setForcedGOGCParams(&ctx)
+			readGlobalGoroutineLeakDetectionParams(&ctx)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/watcher/watcher.go:539]: Don't you also need this on line 517?

File: pkg/pillar/docs/watcher.md (hunk 5 above)
Comment (OhmSpectator) [reply pkg/pillar/docs/watcher.md:10]: It would be nice, but I needed to study this part more to document it in the scope of this PR...

File: pkg/pillar/cmd/watcher/watcher.go (hunk 6 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher.go:539]: I'm thinking now if a more proper place for this is in `handleGlobalConfigImpl`...

Comment (eriknordmark) [reply pkg/pillar/cmd/watcher/watcher.go:539]: Yep.

**Comment by eriknordmark** [comment]
You have a yetus issue as well 
yetus: pkg/pillar/docs/watcher.md#L92
markdownlint:MD047/single-trailing-newline Files should end with a single newline character

**Comment by OhmSpectator** [comment]
> You have a yetus issue as well yetus: pkg/pillar/docs/watcher.md#L92 markdownlint:MD047/single-trailing-newline Files should end with a single newline character

yes, I see, will fix soon

File: pkg/pillar/cmd/watcher/watcher.go [hunk 7]
```diff
@@ -156,6 +243,180 @@ func handleMemoryPressureEvents() {
...
+					// Capacity is sufficient; update entriesToKeep
+					log.Functionf("Capacity sufficient; update entriesToKeep: %d", newEntriesToKeep)
+					entriesToKeep = newEntriesToKeep
+				} else {
+					// Capacity insufficient; create a new slice
+					log.Functionf("Capacity insufficient; create a new slice: %d", newEntriesToKeep+1)
+					newStats := make([]int, len(stats), newEntriesToKeep+1)
+					copy(newStats, stats)
+					stats = newStats
+					entriesToKeep = newEntriesToKeep
```
Comment (christoph-zededa) [inline pkg/pillar/cmd/watcher/watcher.go:362]: Why do you need to create a new slice?

Later you use `append` anyways.

Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher.go:362]: It's a slice of new capacity. It's either bigger or less than the original.

File: pkg/pillar/cmd/watcher/watcher.go [hunk 8]
```diff
@@ -156,6 +243,180 @@ func handleMemoryPressureEvents() {
...
+	// Get the initial goroutine leak detection parameters to create the stats slice
+	goroutinesThreshold, checkInterval, checkStatsFor, keepStatsFor, cooldownPeriod := ctx.GRLDParams.Get()
+	entriesToKeep := int(keepStatsFor / checkInterval)
+	stats := make([]int, 0, entriesToKeep+1)
+	var lastLeakHandled time.Time
+	for {
+		goroutinesThreshold, checkInterval, checkStatsFor, keepStatsFor, cooldownPeriod = ctx.GRLDParams.Get()
+		// Check if we have to resize the stats slice
+		newEntriesToKeep := int(keepStatsFor / checkInterval)
+		if newEntriesToKeep != entriesToKeep {
```
Comment (christoph-zededa) [inline pkg/pillar/cmd/watcher/watcher.go:349]:
```suggestion
		
```
and then later
```
} else if newEntriesToKeep < entriesToKeep {
```

Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher.go:349]: I also need to handle the case of a bigger new capacity.

File: pkg/pillar/cmd/watcher/watcher.go (hunk 7 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher.go:362]: I even have unit tests for both cases!

File: pkg/pillar/cmd/watcher/watcher.go [hunk 9]
```diff
@@ -156,6 +243,180 @@ func handleMemoryPressureEvents() {
...
+		}
+
+		// If we have enough data, detect goroutine leaks
+		if len(stats) > entriesToCheck {
+			// Analyze the data for the last check window
+			entriesInLastCheckWindow := stats[len(stats)-entriesToCheck:]
+			leakDetected, _ := detectGoroutineLeaks(entriesInLastCheckWindow)
+			if leakDetected {
+				// Count the number of goroutines that were created in the last check window
+				numGoroutinesCheckWindowAgo := stats[len(stats)-entriesToCheck]
```
Comment (christoph-zededa) [inline pkg/pillar/cmd/watcher/watcher.go:403]: isn't this just:
```suggestion
				numGoroutinesCheckWindowAgo := entriesInLastCheckWindow[0]
```

File: pkg/pillar/cmd/watcher/watcher.go (hunk 6 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher.go]: Done

File: pkg/pillar/cmd/watcher/watcher.go (hunk 7 above)
Comment (christoph-zededa) [reply pkg/pillar/cmd/watcher/watcher.go:362]: But why do you need a bigger slice if you make it bigger with `append` anyways?

File: pkg/pillar/cmd/watcher/watcher.go (hunk 9 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher.go:403]: Currently, when we use only the last window for the check — yes. However, I have considered a more complex analysis that would incorporate some earlier data. I even implemented this idea but was halted by the need to add a new library, which prevented me from committing the changes. Therefore, I would prefer to keep the current approach for now. I would prefer the analysis slice and the info I use for debug printing to be independent.

File: pkg/pillar/cmd/watcher/watcher.go (hunk 8 above)
Comment (christoph-zededa) [reply pkg/pillar/cmd/watcher/watcher.go:349]: Instead of doing:
```
if a != b {
    if a > b {
    } else {
     }
}
```
I would suggest to do:
```
if a > b {
} else if a < b {
}
```

File: pkg/pillar/cmd/watcher/watcher_test.go [hunk 10]
```diff
@@ -0,0 +1,464 @@
...
+	ctx.GRLDParams.Set(goroutinesThreshold, checkInterval, checkStatsFor, keepStatsFor, cooldownPeriod)
+
+	go goroutinesMonitor(ctx)
+
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+
+	// Create a goroutine to simulate the creation of goroutines
+	go func() {
```
Comment (christoph-zededa) [inline pkg/pillar/cmd/watcher/watcher_test.go:260]: I don't understand why you need this goroutine. Is it to ensure that at least `n` goroutines are running?

File: pkg/pillar/cmd/watcher/watcher.go (hunk 7 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher.go:362]: Mmm... When the capacity is insufficient, does appending cause an implicit allocation?

File: pkg/pillar/cmd/watcher/watcher.go (hunk 8 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher.go:349]: A, I got it... I don't know... The first approach is more readable, IMO.

File: pkg/pillar/cmd/watcher/watcher.go (hunk 9 above)
Comment (christoph-zededa) [reply pkg/pillar/cmd/watcher/watcher.go:403]: Ok.

File: pkg/pillar/cmd/watcher/watcher.go (hunk 7 above)
Comment (christoph-zededa) [reply pkg/pillar/cmd/watcher/watcher.go:362]: Yes

File: pkg/pillar/cmd/watcher/watcher_test.go (hunk 10 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher_test.go:260]: To start creating goroutines in the background... Hmmm... You mean I can just replace it with a loop and wait in the end?

Go has cursed me, and I'm starting to write code with unnecessary goroutines and defers. The next step will be adding channels instead of regular variables =D

File: pkg/pillar/cmd/watcher/watcher.go (hunk 7 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher.go:362]: Ok, I'll simplify the code. Thanks for pointing to it!

Comment (christoph-zededa) [reply pkg/pillar/cmd/watcher/watcher.go:362]: https://go.dev/play/p/3L14ni03ioZ

Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher.go:362]: Yeah, I got it. The only minor concern then is that it can allocate too much

https://go.dev/play/p/czCzIfi1QIs

But who cares about mem in Go, right? =))

File: pkg/pillar/cmd/watcher/watcher_test.go (hunk 10 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher_test.go]: Fixed

File: pkg/pillar/cmd/watcher/watcher.go (hunk 8 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher.go:349]: The code is refactored

File: pkg/pillar/cmd/watcher/watcher.go (hunk 7 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher.go]: Done

File: pkg/pillar/cmd/watcher/watcher_test.go [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+	backupOut := logger.Out
+	// Create a pipe to capture log output
+	r, w, _ := os.Pipe()
+	logger.Out = w
+
+	// Create context with default parameters
+	ctx := &watcherContext{}
+	ctx.GRLDParams.Set(goroutinesThreshold, checkInterval, checkStatsFor, keepStatsFor, cooldownPeriod)
+
+	go goroutinesMonitor(ctx)
```
Comment (christoph-zededa) [inline pkg/pillar/cmd/watcher/watcher_test.go:252]: How do you stop the monitor?

Otherwise you have several of them running in parallel. Or it just does not matter?

Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher_test.go:252]: Good point. From one perspective, running two monitors does not break the test; from another... I was not supposed to do so.

I don't want to add a Context to the function, as it's not needed in the primary use case (running as part of Zedbox). Can I stop the goroutine without adding Context?...

Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher_test.go:252]: image
me, writing Go code and tests

Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher_test.go:252]: > Can I stop the goroutine without adding Context?...

No, I cannot stop a grouting without sharing info between it and the caller...
Will add a new commit soon =\

Comment (OhmSpectator) [reply pkg/pillar/cmd/watcher/watcher_test.go:253]: Done

**Comment by OhmSpectator** [comment]
Fixed races for the logger in the tests

**Review by christoph-zededa — APPROVED** [review]

**Comment by OhmSpectator** [comment]
Suppressed output in the tests, where unnecessary

**Review by eriknordmark — COMMENTED** [review]
Does anything use the Stop() function?
Is the intent that setting some ConfigItems to a value of zero will cause a stop and setting that value back to non-zero will cause a Start()?

**Review by eriknordmark — APPROVED** [review]
Re-run tests

**Comment by OhmSpectator** [comment]
> Does anything use the Stop() function? Is the intent that setting some ConfigItems to a value of zero will cause a stop and setting that value back to non-zero will cause a Start()?

It's used in the tests only, not to run the monitor several times in one package. See https://github.com/lf-edge/eve/pull/4416#discussion_r1834213570

**Comment by OhmSpectator** [comment]
Added the stacks dump file into collect-info if exists.
